### PR TITLE
Change constraint and normalization structure

### PR DESCRIPTION
**Pull request by OliverSchw**

### CODE REVIEW

File: examples/example_gymwrapper.ipynb
Comment (hvater) [inline examples/example_gymwrapper.ipynb:1]: please remove cell outputs from the notebook before commiting

Comment (hvater) [reply examples/example_gymwrapper.ipynb:1]: Or alternatively, run it through once so that the outputs are all there and one can directly see in github what the purpose of the notebook is

File: examples/pmsm_example.ipynb
Comment (hvater) [inline examples/pmsm_example.ipynb:1]: see comment on `example_gymwrapper.ipynb`

File: exciting_environments/utils.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import jax_dataclasses as jdc
+
+
+@jdc.pytree_dataclass
+class Normalization:
```
Comment (hvater) [inline exciting_environments/utils.py:5]: Just as a suggestion:

```suggestion
class MinMaxNormalization:
```

To directly differentiate from other normalization techniques, e.g., logarithmic etc. I am not sure if we ever want to add some other form of normalization, but I think it might prevent some confusion.

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 2]
```diff
@@ -305,72 +313,69 @@ def init_state(self, env_properties, rng: chex.PRNGKey = None, vmap_helper=None)
             phys = self.PhysicalState(
                 deflection=0.0,
                 velocity=0.0,
-                theta=jnp.pi,
+                theta=1.0,
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:316]: Why did you change this? Is the physical state now normalized?

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 3]
```diff
@@ -305,72 +313,69 @@ def init_state(self, env_properties, rng: chex.PRNGKey = None, vmap_helper=None)
...
+                velocity=state_norm[1],
+                theta=state_norm[2],
+                omega=state_norm[3],
             )
             key, subkey = jax.random.split(rng)
         additions = None  # self.Optional(something=jnp.zeros(self.batch_size))
         ref = self.PhysicalState(deflection=jnp.nan, velocity=jnp.nan, theta=jnp.nan, omega=jnp.nan)
-        return self.State(physical_state=phys, PRNGKey=subkey, additions=additions, reference=ref)
+        norm_state = self.State(physical_state=phys, PRNGKey=subkey, additions=additions, reference=ref)
+        return self.denormalize_state(norm_state, env_properties)
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:332]: Do I understand correctly that the state is initialized normalized and then denormalized here?

I think then it is okay. I only want to make sure that the state is still made up of the physical quantities

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 4]
```diff
@@ -305,72 +313,69 @@ def init_state(self, env_properties, rng: chex.PRNGKey = None, vmap_helper=None)
...
 
     @partial(jax.jit, static_argnums=0)
     def generate_reward(self, state, action, env_properties):
         """Returns reward for one batch."""
         reward = 0
+        norm_state = self.normalize_state(state, env_properties)
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:338]: okay I like this 
Seems like a good change to me

generally, it seems to simplify the code for rewards and observations etc :)

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 5]
```diff
@@ -380,8 +385,28 @@ def generate_state_from_observation(self, obs, env_properties, key=None):
         ref = self.PhysicalState(deflection=jnp.nan, velocity=jnp.nan, theta=jnp.nan, omega=jnp.nan)
         with jdc.copy_and_mutate(ref, validate=False) as new_ref:
             for name, pos in zip(self.control_state, range(len(self.control_state))):
-                setattr(new_ref, name, obs[4 + pos] * (getattr(physical_constraints, name)).astype(float))
-        return self.State(physical_state=phys, PRNGKey=subkey, additions=additions, reference=new_ref)
+                setattr(new_ref, name, obs[4 + pos])
+        norm_state = self.State(physical_state=phys, PRNGKey=subkey, additions=additions, reference=new_ref)
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:389]: is the control state a duplication of the reference in the state?
How do they differ and is this sensible?

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 6]
```diff
@@ -305,72 +313,69 @@ def init_state(self, env_properties, rng: chex.PRNGKey = None, vmap_helper=None)
...
+                    getattr(norm_state.reference, name),
                 )
             )
         return obs
 
     @partial(jax.jit, static_argnums=0)
     def generate_state_from_observation(self, obs, env_properties, key=None):
         """Generates state from observation for one batch."""
-        physical_constraints = env_properties.physical_constraints
+        physical_normalizations = env_properties.physical_normalizations
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:373]: this line seems redundant, I think you can remove it

am I missing something?

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 7]
```diff
@@ -380,8 +385,28 @@ def generate_state_from_observation(self, obs, env_properties, key=None):
...
+            for field in fields(phys_soft_const):
+                name = field.name
+                setattr(phys_soft_const, name, jnp.nan)
+            # define soft constraints for physical state
+            soft_constr = jax.nn.relu(jnp.abs(getattr(physical_state_norm, "deflection")) - 1.0)
+            setattr(phys_soft_const, "deflection", soft_constr)
+            soft_constr = jax.nn.relu(jnp.abs(getattr(physical_state_norm, "velocity")) - 1.0)
+            setattr(phys_soft_const, "velocity", soft_constr)
+            soft_constr = jax.nn.relu(jnp.abs(getattr(physical_state_norm, "omega")) - 1.0)
+            setattr(phys_soft_const, "omega", soft_constr)
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:405]: There is probably a way to automate this for arbitrary physical states by adding this computation into the for loop above?

File: exciting_environments/classic_core_env.py [hunk 8]
```diff
@@ -177,7 +177,9 @@ def sim_ahead(self, init_state, actions, env_properties, obs_stepsize, action_st
         )
 
         # denormalize actions
-        actions = actions * jnp.array(tree_flatten(env_properties.action_constraints)[0]).T
+        # TODO for new denormalize_action
+        # actions = actions * jnp.array(tree_flatten(env_properties.action_normalizations)[0]).T
```
Comment (hvater) [inline exciting_environments/classic_core_env.py:181]: Is there still a TODO here?

Is the issue that multidimensional actions are not properly reflected in the current implementation?

File: exciting_environments/core_env.py [hunk 9]
```diff
@@ -69,13 +69,75 @@ def create_in_axes_dataclass(self, dataclass):
...
+                    2
+                    * (
+                        getattr(norm_state.physical_state, name)
+                        - getattr(physical_normalizations, name).min.astype(float)
+                    )
+                    / (
+                        getattr(physical_normalizations, name).max.astype(float)
+                        - getattr(physical_normalizations, name).min.astype(float)
+                    )
+                    - 1
```
Comment (hvater) [inline exciting_environments/core_env.py:99]: just as an idea. Is there a point in moving some of this implementation into the (MinMax)-Normalization class?

i think passing getattr(norm_state.physical_state, name) to its normalization could do the same as this?

We can discuss this if you want

File: exciting_environments/mass_spring_damper/mass_spring_damper_env.py [hunk 10]
```diff
@@ -285,8 +293,23 @@ def generate_state_from_observation(self, obs, env_properties, key=None):
...
+
+    def default_soft_constraints(self, state, action_norm, env_properties):
+        state_norm = self.normalize_state(state, env_properties)
+        physical_state_norm = state_norm.physical_state
+        with jdc.copy_and_mutate(physical_state_norm, validate=False) as phys_soft_const:
+            # define soft constraints for physical state
+            soft_constr = jax.nn.relu(jnp.abs(getattr(physical_state_norm, "deflection")) - 1.0)
+            setattr(phys_soft_const, "deflection", soft_constr)
+            soft_constr = jax.nn.relu(jnp.abs(getattr(physical_state_norm, "velocity")) - 1.0)
+            setattr(phys_soft_const, "velocity", soft_constr)
```
Comment (hvater) [inline exciting_environments/mass_spring_damper/mass_spring_damper_env.py:308]: Automation for arbitrary physical states could be possible? See cart pole comments

File: exciting_environments/pmsm/pmsm_env.py [hunk 11]
```diff
@@ -365,7 +374,8 @@ def currents_to_torque(self, i_d, i_q, env_properties):
     def currents_to_torque_saturated(self, i_d, i_q, env_properties):
         Psi_d = self.LUT_interpolators["Psi_d"](jnp.array([i_d, i_q]))
         Psi_q = self.LUT_interpolators["Psi_q"](jnp.array([i_d, i_q]))
-        return (Psi_d * i_q - Psi_q * i_d)[0]
+        t = 3 / 2 * env_properties.static_params.p * (Psi_d * i_q - Psi_q * i_d)[0]
+        return t
 
     def init_state(self, env_properties, rng: chex.PRNGKey = None, vmap_helper=None):
```
Comment (hvater) [inline exciting_environments/pmsm/pmsm_env.py:380]: For the other systems you defined the state in normalized form and then denormalized it.

Is there a reason you did this differently for the PMSM?

I think you could simplify the initial values for i_d and omega_el this way (should both be 0 in normalized form?)

**Review by hvater — COMMENTED** [review]
Overall the changes look good and I think it cleans some parts up nicely.

Just have a look at my comments in work in what you think is important.
There is nothing that I see to be a major problem.

File: exciting_environments/utils.py (hunk 1 above)
Comment (OliverSchw) [reply exciting_environments/utils.py:5]: Sounds good to me.

File: exciting_environments/cart_pole/cart_pole_env.py [hunk 12]
```diff
@@ -57,36 +60,40 @@ def __init__(
...
         Args:
             batch_size(int): Number of training examples utilized in one iteration. Default: 8
-            physical_constraints(dict): Constraints of physical state of the environment.
-                deflection(float): Deflection of the cart. Default: 10
-                velocity(float): Velocity of the cart. Default: 10
-                theta(float): Rotation angle of the pole. Default: jnp.pi
-                omega(float): Angular velocity. Default: 10
-            action_constraints(dict): Constraints of the action.
-                force(float): Maximum torque that can be applied to the system as action. Default: 20
+            physical_normalizations(dict): min-max normalization values of the physical state of the environment.
```
Comment (hvater) [inline exciting_environments/cart_pole/cart_pole_env.py:63]: _Non-blocking_

maybe we should specify the units here or at least we should note somewhere that all physical properties are given in SI units.

File: exciting_environments/core_env.py [hunk 13]
```diff
@@ -53,6 +53,202 @@ def __init__(
...
+        self.physical_state_dim = len(fields(self.PhysicalState))
+
+    @abstractmethod
+    @jdc.pytree_dataclass
+    class PhysicalState:
+        """The physical state x(t) of the underlying system and whose derivative
+        w.r.t. time is described in the underlying ODE.
+
+        The values stored in this dataclass are expected to be actual physical values
+        that are unnormalized.
```
Comment (hvater) [inline exciting_environments/core_env.py:66]: _Non-blocking_

, i.e., the values are all given in SI units.

File: exciting_environments/pmsm/pmsm_env.py [hunk 14]
```diff
@@ -287,22 +295,22 @@ class Action:
         u_d: jax.Array
         u_q: jax.Array
 
-    @jdc.pytree_dataclass
-    class State:
-        """Dataclass used for simulation which contains environment specific dataclasses."""
+    # @jdc.pytree_dataclass
```
Comment (hvater) [inline exciting_environments/pmsm/pmsm_env.py:298]: delete if not needed anymore

**Review by hvater — APPROVED** [review]
Looks good

Only minimal comments that you can consider. Nothing detrimental
